core/src/main/java/com/linecorp/armeria/client/HttpRequestSubscriber.java

```diff
@@ -203,6 +203,7 @@ private void writeFirstHeader(HttpSession session) {
         logBuilder.requestHeaders(merged);
         final ChannelFuture future = encoder.writeHeaders(id, streamId(), merged, isEmpty);
```

Comment:
- What if we catch the exception here using `try-catch` and notify the listener by ourselves? Do we still need the `INIT` state?
  ```java
  ChannelFuture future;
  try {
      ChannelFuture future = encoder.writeHeaders(...);
      future.addListener(this);
      ch.flush();
  } catch (HeaderListSizeException e) {
      // Notify the exception later.
      ch.eventLoop().execute(() -> ch.newFailedFuture(e).addListener(this));
  }
  ```
  - We could probably modify `encoder.writeHeaders()` by moving this `try-catch` inside?
- In some cases, the `future` returned by `writeHeaders()` may be complete even before we call `addListener(this)`, making the listener (`this.operationComplete()`) invoked immediately and directly. Would this be a problem as well?

Reply:
> try-catch and notify the listener by ourselves? 

`encoder.writeHeaders` did not raise a `HeaderListSizeException` was propagated to `onStreamClosed()` and the returned `ChannelFuture`.
So I don't see we can fix the problem with `try-catch`.

> making the listener (this.operationComplete()) invoked immediately and directly. Would this be a problem as well?

If `HeaderListSizeException` is raised, `onStreamClosed` is called first and `encoder.writeHeaders(...)` returns the failed future.
While `onStreamClosed` is being called, the response is closed with `ClosedStreamException`. Therefore, we can't close a response with the notified cause from `operationComplete()`.

Reply:
FYI, this is the stack trace for `HeaderListSizeException`
```
com.linecorp.armeria.common.stream.ClosedStreamException: null
	at com.linecorp.armeria.common.stream.ClosedStreamException.get(ClosedStreamException.java:36)
	at com.linecorp.armeria.client.Http2ResponseDecoder.onStreamClosed(Http2ResponseDecoder.java:144)
	at io.netty.handler.codec.http2.DefaultHttp2Connection.notifyClosed(DefaultHttp2Connection.java:357)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$ActiveStreams.removeFromActiveStreams(DefaultHttp2Connection.java:1007)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$ActiveStreams.deactivate(DefaultHttp2Connection.java:963)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$DefaultStream.close(DefaultHttp2Connection.java:515)
	at io.netty.handler.codec.http2.DefaultHttp2Connection$DefaultStream.close(DefaultHttp2Connection.java:521)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.closeStream(Http2ConnectionHandler.java:613)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.processRstStreamWriteResult(Http2ConnectionHandler.java:894)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.resetStream(Http2ConnectionHandler.java:804)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.onStreamError(Http2ConnectionHandler.java:729)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.onError(Http2ConnectionHandler.java:634)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder.writeHeaders0(DefaultHttp2ConnectionEncoder.java:251)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder.writeHeaders(DefaultHttp2ConnectionEncoder.java:151)
	at com.linecorp.armeria.client.ClientHttp2ObjectEncoder.doWriteHeaders(ClientHttp2ObjectEncoder.java:70)
	at com.linecorp.armeria.client.ClientHttpObjectEncoder.writeHeaders(ClientHttpObjectEncoder.java:41)
	at com.linecorp.armeria.client.HttpRequestSubscriber.writeFirstHeader(HttpRequestSubscriber.java:204)
	at com.linecorp.armeria.client.HttpRequestSubscriber.onSubscribe(HttpRequestSubscriber.java:180)
```

Reply:
By the way, I realized that we can pass an early created `ChannelPromise` as an argument.
The `INIT` state is now removed.